// @yaga/leaflet-ng2/lib/image-overlay.directive.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var leaflet_1 = require("leaflet");
var consts_1 = require("./consts");
var layer_group_provider_1 = require("./layer-group.provider");
var layer_provider_1 = require("./layer.provider");
/**
 * Angular2 directive for Leaflet image overlays.
 *
 * *You can use this directive in an Angular2 template after importing `YagaModule`.*
 *
 * How to use in a template:
 * ```html
 * <yaga-map>
 *     <yaga-image-overlay
 *         [(url)]="..."
 *         [(display)]="..."
 *         [(opacity)]="..."
 *         [(zIndex)]="..."
 *         [(bounds)]="..."
 *         [(north)]="..."
 *         [(east)]="..."
 *         [(south)]="..."
 *         [(west)]="..."
 *
 *         (add)="..."
 *         (remove)="..."
 *         (popupopen)="..."
 *         (popupclose)="..."
 *         (tooltipopen)="..."
 *         (tooltipclose)="..."
 *         (click)="..."
 *         (dblclick)="..."
 *         (mousedown)="..."
 *         (mouseover)="..."
 *         (mouseout)="..."
 *         (contextmenu)="..."
 *         (load)="..."
 *         (error)="..."
 *
 *         [crossOrigin]="..."
 *         [alt]="..."
 *         [interactive]="..."
 *         [attribution]="..."
 *         >
 *     </yaga-image-overlay>
 * </yaga-map>
 * ```
 *
 * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay Original Leaflet documentation
 * @link https://leaflet-ng2.yagajs.org/latest/browser-test?grep=Image-Overlay%20Directive Unit-Test
 * @link https://leaflet-ng2.yagajs.org/latest/coverage/lcov-report/lib/image-overlay.directive.js.html Test coverage
 * @link https://leaflet-ng2.yagajs.org/latest/typedoc/classes/imageoverlaydirective.html API documentation
 * @example https://leaflet-ng2.yagajs.org/latest/examples/image-overlay-directive
 */
var ImageOverlayDirective = /** @class */ (function (_super) {
    __extends(ImageOverlayDirective, _super);
    function ImageOverlayDirective(layerGroupProvider, layerProvider) {
        var _this = 
        // Transparent 1px image:
        _super.call(this, consts_1.TRANSPARENT_PIXEL, [[0, 0], [1, 1]], {}) || this;
        _this.layerGroupProvider = layerGroupProvider;
        /**
         * Two-Way bound property for the URL.
         * Use it with `<yaga-image-overlay [(url)]="someValue">` or
         * `<yaga-image-overlay (urlChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-seturl Original Leaflet documentation
         */
        _this.urlChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the display status of the layer.
         * Use it with `<yaga-image-overlay [(display)]="someValue">`
         * or `<yaga-image-overlay (displayChange)="processEvent($event)">`
         */
        _this.displayChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the opacity of the layer.
         * Use it with `<yaga-image-overlay [(opacity)]="someValue">`
         * or `<yaga-image-overlay (opacityChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-opacity Original Leaflet documentation
         */
        _this.opacityChange = new core_1.EventEmitter();
        // maybe implement -> @Output() public zIndexChange: EventEmitter<number> = new EventEmitter();
        /**
         * Two-Way bound property for the bounds of the image.
         * Use it with `<yaga-image-overlay [(bounds)]="someValue">`
         * or `<yaga-image-overlay (opacityChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
         */
        _this.boundsChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the north bounds of the image.
         * Use it with `<yaga-image-overlay [(north)]="someValue">`
         * or `<yaga-image-overlay (northChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
         */
        _this.northChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the east bounds of the image.
         * Use it with `<yaga-image-overlay [(east)]="someValue">`
         * or `<yaga-image-overlay (eastChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
         */
        _this.eastChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the south bounds of the image.
         * Use it with `<yaga-image-overlay [(south)]="someValue">`
         * or `<yaga-image-overlay (southChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
         */
        _this.southChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the west bounds of the image.
         * Use it with `<yaga-image-overlay [(west)]="someValue">`
         * or `<yaga-image-overlay (westChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
         */
        _this.westChange = new core_1.EventEmitter();
        /**
         * From leaflet fired add event.
         * Use it with `<yaga-image-overlay (add)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-add Original Leaflet documentation
         */
        _this.addEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired remove event.
         * Use it with `<yaga-image-overlay (remove)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-remove Original Leaflet documentation
         */
        _this.removeEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired popupopen event.
         * Use it with `<yaga-image-overlay (popupopen)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-popupopen Original Leaflet documentation
         */
        _this.popupopenEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired popupclose event.
         * Use it with `<yaga-image-overlay (popupclose)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-popupclose Original Leaflet documentation
         */
        _this.popupcloseEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired tooltipopen event.
         * Use it with `<yaga-image-overlay (tooltipopen)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-tooltipopen Original Leaflet documentation
         */
        _this.tooltipopenEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired tooltipclose event.
         * Use it with `<yaga-image-overlay (tooltipclose)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-tooltipclose Original Leaflet documentation
         */
        _this.tooltipcloseEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired click event.
         * Use it with `<yaga-image-overlay (click)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-click Original Leaflet documentation
         */
        _this.clickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired dblclick event.
         * Use it with `<yaga-image-overlay (dblclick)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-dblclick Original Leaflet documentation
         */
        _this.dblclickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mousedown event.
         * Use it with `<yaga-image-overlay (mousedown)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-mousedown Original Leaflet documentation
         */
        _this.mousedownEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseover event.
         * Use it with `<yaga-image-overlay (mouseover)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-mouseover Original Leaflet documentation
         */
        _this.mouseoverEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseout event.
         * Use it with `<yaga-image-overlay (mouseout)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-mouseout Original Leaflet documentation
         */
        _this.mouseoutEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired contextmenu event.
         * Use it with `<yaga-image-overlay (contextmenu)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-contextmenu Original Leaflet documentation
         */
        _this.contextmenuEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired load event.
         * Use it with `<yaga-image-overlay (load)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-load Original Leaflet documentation
         */
        _this.loadEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired error event.
         * Use it with `<yaga-image-overlay (error)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-error Original Leaflet documentation
         */
        _this.errorEvent = new core_1.EventEmitter();
        layerProvider.ref = _this;
        _this.on("remove", function () {
            _this.displayChange.emit(false);
        });
        _this.on("add", function () {
            _this.displayChange.emit(true);
        });
        _this.layerGroupProvider.ref.addLayer(_this);
        // Events
        _this.on("add", function (event) {
            _this.addEvent.emit(event);
        });
        _this.on("remove", function (event) {
            _this.removeEvent.emit(event);
        });
        _this.on("popupopen", function (event) {
            _this.popupopenEvent.emit(event);
        });
        _this.on("popupclose", function (event) {
            _this.popupcloseEvent.emit(event);
        });
        _this.on("tooltipopen", function (event) {
            _this.tooltipopenEvent.emit(event);
        });
        _this.on("tooltipclose", function (event) {
            _this.tooltipcloseEvent.emit(event);
        });
        _this.on("click", function (event) {
            _this.clickEvent.emit(event);
        });
        _this.on("dblclick", function (event) {
            _this.dblclickEvent.emit(event);
        });
        _this.on("mousedown", function (event) {
            _this.mousedownEvent.emit(event);
        });
        _this.on("mouseover", function (event) {
            _this.mouseoverEvent.emit(event);
        });
        _this.on("mouseout", function (event) {
            _this.mouseoutEvent.emit(event);
        });
        _this.on("contextmenu", function (event) {
            _this.contextmenuEvent.emit(event);
        });
        _this.on("load", function (event) {
            _this.loadEvent.emit(event);
        });
        _this.on("error", function (event) {
            _this.errorEvent.emit(event);
        });
        return _this;
    }
    /**
     * This function gets called from Angular on destroy of the html-component.
     * @link https://angular.io/docs/ts/latest/api/core/index/OnDestroy-class.html
     */
    ImageOverlayDirective.prototype.ngOnDestroy = function () {
        this.removeFrom(this.layerGroupProvider.ref);
    };
    /**
     * Derived method of the original setUrl method.
     * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-seturl Original Leaflet documentation
     */
    ImageOverlayDirective.prototype.setUrl = function (url) {
        if (this.url === url) {
            return this;
        }
        this.urlChange.emit(url);
        return _super.prototype.setUrl.call(this, url);
    };
    Object.defineProperty(ImageOverlayDirective.prototype, "url", {
        get: function () {
            return this._url;
        },
        /**
         * Two-Way bound property for the URL.
         * Use it with `<yaga-image-overlay [(url)]="someValue">` or `<yaga-image-overlay [url]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-seturl Original Leaflet documentation
         */
        set: function (val) {
            this.setUrl(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Derived method of the original setOpacity method.
     * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setopacity Original Leaflet documentation
     */
    ImageOverlayDirective.prototype.setOpacity = function (val) {
        if (this.opacity === val) {
            return this;
        }
        this.opacityChange.emit(val);
        return _super.prototype.setOpacity.call(this, val);
    };
    Object.defineProperty(ImageOverlayDirective.prototype, "opacity", {
        get: function () {
            if (this.options.hasOwnProperty("opacity") && this.options.opacity !== undefined) {
                return this.options.opacity;
            }
            return 1;
        },
        /**
         * Two-Way bound property for the opacity.
         * Use it with `<yaga-image-overlay [(opacity)]="someValue">` or `<yaga-image-overlay [opacity]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setopacity Original Leaflet documentation
         */
        set: function (val) {
            this.setOpacity(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "display", {
        /**
         * Two-Way bound property for the display status of the layer.
         * Use it with `<yaga-image-overlay [(display)]="someValue">` or `<yaga-image-overlay [display]="someValue">`
         */
        get: function () {
            var pane;
            var container;
            try {
                pane = this.getPane();
                container = this.getElement();
            }
            catch (err) {
                /* istanbul ignore next */
                return false;
            }
            /* tslint:disable:prefer-for-of */
            for (var i = 0; i < pane.children.length; i += 1) {
                /* tslint:enable */
                /* istanbul ignore else */
                if (pane.children[i] === container) {
                    return true;
                }
            }
            return false;
        },
        /**
         * Two-Way bound property for the display status of the layer.
         * Use it with `<yaga-image-overlay [(display)]="someValue">` or `<yaga-image-overlay [display]="someValue">`
         */
        set: function (val) {
            var isDisplayed = this.display;
            if (isDisplayed === val) {
                return;
            }
            var pane;
            var container;
            var map;
            var events; // Dictionary of functions
            var eventKeys;
            try {
                pane = this.getPane();
                container = this.getElement();
                map = this._map;
                events = this.getEvents();
                eventKeys = Object.keys(events);
            }
            catch (err) {
                /* istanbul ignore next */
                return;
            }
            if (val) {
                // show layer
                pane.appendChild(container);
                for (var _i = 0, eventKeys_1 = eventKeys; _i < eventKeys_1.length; _i++) {
                    var eventKey = eventKeys_1[_i];
                    map.on(eventKey, events[eventKey], this);
                }
            }
            else {
                // hide layer
                pane.removeChild(container);
                for (var _a = 0, eventKeys_2 = eventKeys; _a < eventKeys_2.length; _a++) {
                    var eventKey = eventKeys_2[_a];
                    map.off(eventKey, events[eventKey], this);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Derived method of the original setBounds method.
     * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-setbounds Original Leaflet documentation
     */
    ImageOverlayDirective.prototype.setBounds = function (val) {
        _super.prototype.setBounds.call(this, leaflet_1.latLngBounds(val));
        this.boundsChange.emit(this.bounds);
        this.northChange.emit(this.north);
        this.eastChange.emit(this.east);
        this.southChange.emit(this.south);
        this.westChange.emit(this.west);
        return this;
    };
    Object.defineProperty(ImageOverlayDirective.prototype, "bounds", {
        get: function () {
            return this.getBounds();
        },
        /**
         * Two-Way bound property for the bounds of the image.
         * Use it with `<yaga-image-overlay [(bounds)]="someValue">` or `<yaga-image-overlay [bounds]="someValue">`
         */
        set: function (val) {
            this.setBounds(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "north", {
        get: function () {
            return this.getBounds().getNorth();
        },
        /**
         * Two-Way bound property for the north bounds of the image.
         * Use it with `<yaga-image-overlay [(north)]="someValue">` or `<yaga-image-overlay [north]="someValue">`
         */
        set: function (val) {
            var oldBounds = this.getBounds();
            // super because we call the change listeners ourselves
            _super.prototype.setBounds.call(this, leaflet_1.latLngBounds([
                [oldBounds.getSouth(), oldBounds.getWest()],
                [val, oldBounds.getEast()],
            ]));
            this.boundsChange.emit(this.bounds);
            this.northChange.emit(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "east", {
        get: function () {
            return this.getBounds().getEast();
        },
        /**
         * Two-Way bound property for the east bounds of the image.
         * Use it with `<yaga-image-overlay [(east)]="someValue">` or `<yaga-image-overlay [east]="someValue">`
         */
        set: function (val) {
            var oldBounds = this.getBounds();
            _super.prototype.setBounds.call(this, leaflet_1.latLngBounds([
                [oldBounds.getSouth(), oldBounds.getWest()],
                [oldBounds.getNorth(), val],
            ]));
            this.boundsChange.emit(this.bounds);
            this.eastChange.emit(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "south", {
        get: function () {
            return this.getBounds().getSouth();
        },
        /**
         * Two-Way bound property for the south bounds of the image.
         * Use it with `<yaga-image-overlay [(south)]="someValue">` or `<yaga-image-overlay [south]="someValue">`
         */
        set: function (val) {
            var oldBounds = this.getBounds();
            _super.prototype.setBounds.call(this, leaflet_1.latLngBounds([
                [val, oldBounds.getWest()],
                [oldBounds.getNorth(), oldBounds.getEast()],
            ]));
            this.boundsChange.emit(this.bounds);
            this.southChange.emit(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "west", {
        get: function () {
            return this.getBounds().getWest();
        },
        /**
         * Two-Way bound property for the west bounds of the image.
         * Use it with `<yaga-image-overlay [(west)]="someValue">` or `<yaga-image-overlay [west]="someValue">`
         */
        set: function (val) {
            var oldBounds = this.getBounds();
            _super.prototype.setBounds.call(this, leaflet_1.latLngBounds([
                [oldBounds.getSouth(), val],
                [oldBounds.getNorth(), oldBounds.getEast()],
            ]));
            this.boundsChange.emit(this.bounds);
            this.westChange.emit(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "crossOrigin", {
        get: function () {
            return !!this.options.crossOrigin;
        },
        /**
         * Input for the crossOrigin.
         * Use it with `<yaga-image-overlay [crossOrigin]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-crossorigin Original Leaflet documentation
         */
        set: function (val) {
            this.options.crossOrigin = val;
            this.getElement().crossOrigin = val ? "" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "alt", {
        get: function () {
            return this.getElement().alt;
        },
        /**
         * Input for the alternative text.
         * Use it with `<yaga-image-overlay [alt]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-alt Original Leaflet documentation
         */
        set: function (val) {
            this.options.alt = val;
            this.getElement().alt = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "interactive", {
        get: function () {
            return !!this.options.interactive;
        },
        /**
         * Input for the state of interaction.
         * Use it with `<yaga-image-overlay [interactive]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-interactive Original Leaflet documentation
         */
        set: function (val) {
            this.options.interactive = val;
            this.onRemove(this._map);
            this.onAdd(this._map);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayDirective.prototype, "attribution", {
        get: function () {
            return this.getAttribution() || "";
        },
        /**
         * Input for the attribution.
         * Use it with `<yaga-image-overlay [attribution]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#imageoverlay-attribution Original Leaflet documentation
         */
        set: function (val) {
            if (this._map && this._map.attributionControl) {
                var oldAttribution = this.getAttribution();
                if (oldAttribution) {
                    this._map.attributionControl.removeAttribution(oldAttribution);
                }
                this._map.attributionControl.addAttribution(val);
            }
            this.options.attribution = val;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "urlChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "displayChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "opacityChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "boundsChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "northChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "eastChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "southChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "westChange", void 0);
    __decorate([
        core_1.Output("add"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "addEvent", void 0);
    __decorate([
        core_1.Output("remove"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "removeEvent", void 0);
    __decorate([
        core_1.Output("popupopen"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "popupopenEvent", void 0);
    __decorate([
        core_1.Output("popupclose"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "popupcloseEvent", void 0);
    __decorate([
        core_1.Output("tooltipopen"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "tooltipopenEvent", void 0);
    __decorate([
        core_1.Output("tooltipclose"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "tooltipcloseEvent", void 0);
    __decorate([
        core_1.Output("click"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "clickEvent", void 0);
    __decorate([
        core_1.Output("dblclick"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "dblclickEvent", void 0);
    __decorate([
        core_1.Output("mousedown"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "mousedownEvent", void 0);
    __decorate([
        core_1.Output("mouseover"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "mouseoverEvent", void 0);
    __decorate([
        core_1.Output("mouseout"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "mouseoutEvent", void 0);
    __decorate([
        core_1.Output("contextmenu"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "contextmenuEvent", void 0);
    __decorate([
        core_1.Output("load"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "loadEvent", void 0);
    __decorate([
        core_1.Output("error"),
        __metadata("design:type", core_1.EventEmitter)
    ], ImageOverlayDirective.prototype, "errorEvent", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ImageOverlayDirective.prototype, "url", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], ImageOverlayDirective.prototype, "opacity", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ImageOverlayDirective.prototype, "display", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", leaflet_1.LatLngBounds),
        __metadata("design:paramtypes", [leaflet_1.LatLngBounds])
    ], ImageOverlayDirective.prototype, "bounds", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], ImageOverlayDirective.prototype, "north", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], ImageOverlayDirective.prototype, "east", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], ImageOverlayDirective.prototype, "south", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], ImageOverlayDirective.prototype, "west", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ImageOverlayDirective.prototype, "crossOrigin", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ImageOverlayDirective.prototype, "alt", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ImageOverlayDirective.prototype, "interactive", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ImageOverlayDirective.prototype, "attribution", null);
    ImageOverlayDirective = __decorate([
        core_1.Directive({
            providers: [layer_provider_1.LayerProvider],
            selector: "yaga-image-overlay",
        }),
        __metadata("design:paramtypes", [layer_group_provider_1.LayerGroupProvider,
            layer_provider_1.LayerProvider])
    ], ImageOverlayDirective);
    return ImageOverlayDirective;
}(leaflet_1.ImageOverlay));
exports.ImageOverlayDirective = ImageOverlayDirective;
