// @yaga/leaflet-ng2/lib/map.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var leaflet_1 = require("leaflet");
var consts_1 = require("./consts");
var layer_group_provider_1 = require("./layer-group.provider");
var map_provider_1 = require("./map.provider");
/**
 * Angular2 root component for a Leaflet map
 *
 * *You can use this directive in an Angular2 template after importing `YagaModule`.*
 *
 * How to use in a template:
 * ```html
 * <yaga-map
 *     [(zoom)]="..."
 *     [(lat)]="..."
 *     [(lng)]="..."
 *     [(minZoom)]="..."
 *     [(maxZoom)]="..."
 *     [(maxBounds)]="..."
 *
 *     (baselayerchange)="..."
 *     (overlayadd)="..."
 *     (overlayremove)="..."
 *     (layeradd)="..."
 *     (layerremove)="..."
 *     (zoomlevelschan)="..."
 *     (resize)="..."
 *     (unload)="..."
 *     (viewreset)="..."
 *     (load)="..."
 *     (zoomstart)="..."
 *     (movestart)="..."
 *     (zoom)="..."
 *     (move)="..."
 *     (zoomend)="..."
 *     (moveend)="..."
 *     (popupopen)="..."
 *     (popupclose)="..."
 *     (autopanstart)="..."
 *     (tooltipopen)="..."
 *     (tooltipclose)="..."
 *     (click)="..."
 *     (dblclick)="..."
 *     (mousedown)="..."
 *     (mouseup)="..."
 *     (mouseover)="..."
 *     (mouseout)="..."
 *     (mousemove)="..."
 *     (contextmenu)="..."
 *     (keypress)="..."
 *     (preclick)="..."
 *     (zoomanim)="..."
 *
 *     [crs]="..."
 *     [closePopupOnClick]="..."
 *     [zoomSnap]="..."
 *     [zoomDelta]="..."
 *     [trackResize]="..."
 *     [boxZoomEnabled]="..."
 *     [doubleClickZoomEnabled]="..."
 *     [draggingEnabled]="..."
 *     [fadeAnimation]="..."
 *     [markerZoomAnimation]="..."
 *     [transform3DLimit]="..."
 *     [zoomAnimation]="..."
 *     [zoomAnimationThreshold]="..."
 *     [inertia]="..."
 *     [inertiaDeceleration]="..."
 *     [inertiaMaxSpeed]="..."
 *     [easeLinearity]="..."
 *     [worldCopyJump]="..."
 *     [maxBoundsViscosity]="..."
 *     [keyboardEnabled]="..."
 *     [keyboardPanDelta]="..."
 *     [scrollWheelZoomEnabled]="..."
 *     [wheelDebounceTime]="..."
 *     [wheelPxPerZoomLevel]="..."
 *     [tapEnabled]="..."
 *     [tapTolerance]="..."
 *     [bounceAtZoomLimits]="..."
 *     [touchZoomEnabled]="...">
 *     <!-- other yaga directives -->
 * </yaga-map>
 * ```
 *
 * You can use the following directives as child of this one:
 *
 * * yaga-attribution-control
 * * yaga-circle
 * * yaga-circle-marker
 * * yaga-geojson
 * * yaga-image-overlay
 * * yaga-marker
 * * yaga-polygon
 * * yaga-polyline
 * * yaga-rectangle
 * * yaga-scale-control
 * * yaga-tile-layer
 * * yaga-wms-layer
 * * yaga-zoom-control
 *
 * @link http://leafletjs.com/reference-1.2.0.html#tilelayer Original Leaflet documentation
 * @link https://leaflet-ng2.yagajs.org/latest/browser-test?grep=Tile-Layer%20Directive Unit-Test
 * @link https://leaflet-ng2.yagajs.org/latest/coverage/lcov-report/lib/tile-layer.directive.js.html Test coverage
 * @link https://leaflet-ng2.yagajs.org/latest/typedoc/classes/tilelayerdirective.html API documentation
 * @example https://leaflet-ng2.yagajs.org/latest/examples/tile-layer-directive
 */
var MapComponent = /** @class */ (function (_super) {
    __extends(MapComponent, _super);
    function MapComponent(elementRef, layerProvider, mapProvider) {
        var _this = _super.call(this, elementRef.nativeElement, { attributionControl: false, zoomControl: false }) || this;
        /**
         * Two-Way bound property for the zoom.
         * Use it with `<yaga-map [(zoom)]="someValue">` or `<yaga-map (zoomChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setzoom Original Leaflet documentation
         */
        _this.zoomChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the center latitude.
         * Use it with `<yaga-map [(lat)]="someValue">` or `<yaga-map (latChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setview Original Leaflet documentation
         */
        _this.latChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the center longitude.
         * Use it with `<yaga-map [(lng)]="someValue">` or `<yaga-map (lngChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setview Original Leaflet documentation
         */
        _this.lngChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the minimal available zoom.
         * Use it with `<yaga-map [(minZoom)]="someValue">` or `<yaga-map (minZoomChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setminzoom Original Leaflet documentation
         */
        _this.minZoomChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the maximal available zoom.
         * Use it with `<yaga-map [(maxZoom)]="someValue">` or `<yaga-map (maxZoomChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setmaxzoom Original Leaflet documentation
         */
        _this.maxZoomChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the bounds on the map.
         * Use it with `<yaga-map [(maxBounds)]="someValue">`
         * or `<yaga-map (maxBoundsChange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setmaxbounds Original Leaflet documentation
         */
        _this.maxBoundsChange = new core_1.EventEmitter();
        /**
         * From leaflet fired baselayerchange event.
         * Use it with `<yaga-tile-layer (baselayerchange)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-baselayerchange Original Leaflet documentation
         */
        _this.baselayerchangeEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired overlayadd event.
         * Use it with `<yaga-tile-layer (overlayadd)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-overlayadd Original Leaflet documentation
         */
        _this.overlayaddEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired overlayremove event.
         * Use it with `<yaga-tile-layer (overlayremove)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-overlayremove Original Leaflet documentation
         */
        _this.overlayremoveEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired layeradd event.
         * Use it with `<yaga-tile-layer (layeradd)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-layeradd Original Leaflet documentation
         */
        _this.layeraddEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired layerremove event.
         * Use it with `<yaga-tile-layer (layerremove)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-layerremove Original Leaflet documentation
         */
        _this.layerremoveEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired zoomlevelschan event.
         * Use it with `<yaga-tile-layer (zoomlevelschan)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomlevelschan Original Leaflet documentation
         */
        _this.zoomlevelschangeEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired resize event.
         * Use it with `<yaga-tile-layer (resize)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-resize Original Leaflet documentation
         */
        _this.resizeEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired unload event.
         * Use it with `<yaga-tile-layer (unload)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-unload Original Leaflet documentation
         */
        _this.unloadEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired viewreset event.
         * Use it with `<yaga-tile-layer (viewreset)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-viewreset Original Leaflet documentation
         */
        _this.viewresetEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired load event.
         * Use it with `<yaga-tile-layer (load)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-load Original Leaflet documentation
         */
        _this.loadEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired zoomstart event.
         * Use it with `<yaga-tile-layer (zoomstart)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomstart Original Leaflet documentation
         */
        _this.zoomstartEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired movestart event.
         * Use it with `<yaga-tile-layer (movestart)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-movestart Original Leaflet documentation
         */
        _this.movestartEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired zoom event.
         * Use it with `<yaga-tile-layer (zoom)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoom Original Leaflet documentation
         */
        _this.zoomEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired move event.
         * Use it with `<yaga-tile-layer (move)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-move Original Leaflet documentation
         */
        _this.moveEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired zoomend event.
         * Use it with `<yaga-tile-layer (zoomend)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomend Original Leaflet documentation
         */
        _this.zoomendEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired moveend event.
         * Use it with `<yaga-tile-layer (moveend)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-moveend Original Leaflet documentation
         */
        _this.moveendEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired (undocumented) boxzoomstart event.
         * Use it with `<yaga-tile-layer (boxzoomstart)="processEvent($event)">`
         * @link https://github.com/Leaflet/Leaflet/blob/master/src/map/handler/Map.BoxZoom.js SourceCode
         */
        _this.boxzoomstartEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired (undocumented) boxzoomend event.
         * Use it with `<yaga-tile-layer (boxzoomend)="processEvent($event)">`
         * @link https://github.com/Leaflet/Leaflet/blob/master/src/map/handler/Map.BoxZoom.js SourceCode
         */
        _this.boxzoomendEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired popupopen event.
         * Use it with `<yaga-tile-layer (popupopen)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-popupopen Original Leaflet documentation
         */
        _this.popupopenEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired popupclose event.
         * Use it with `<yaga-tile-layer (popupclose)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-popupclose Original Leaflet documentation
         */
        _this.popupcloseEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired autopanstart event.
         * Use it with `<yaga-tile-layer (autopanstart)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-autopanstart Original Leaflet documentation
         */
        _this.autopanstartEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired tooltipopen event.
         * Use it with `<yaga-tile-layer (tooltipopen)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-tooltipopen Original Leaflet documentation
         */
        _this.tooltipopenEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired tooltipclose event.
         * Use it with `<yaga-tile-layer (tooltipclose)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-tooltipclose Original Leaflet documentation
         */
        _this.tooltipcloseEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired click event.
         * Use it with `<yaga-tile-layer (click)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-click Original Leaflet documentation
         */
        _this.clickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired dblclick event.
         * Use it with `<yaga-tile-layer (dblclick)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-dblclick Original Leaflet documentation
         */
        _this.dblclickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mousedown event.
         * Use it with `<yaga-tile-layer (mousedown)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-mousedown Original Leaflet documentation
         */
        _this.mousedownEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseup event.
         * Use it with `<yaga-tile-layer (mouseup)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-mouseup Original Leaflet documentation
         */
        _this.mouseupEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseover event.
         * Use it with `<yaga-tile-layer (mouseover)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-mouseover Original Leaflet documentation
         */
        _this.mouseoverEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseout event.
         * Use it with `<yaga-tile-layer (mouseout)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-mouseout Original Leaflet documentation
         */
        _this.mouseoutEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mousemove event.
         * Use it with `<yaga-tile-layer (mousemove)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-mousemove Original Leaflet documentation
         */
        _this.mousemoveEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired contextmenu event.
         * Use it with `<yaga-tile-layer (contextmenu)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-contextmenu Original Leaflet documentation
         */
        _this.contextmenuEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired keypress event.
         * Use it with `<yaga-tile-layer (keypress)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-keypress Original Leaflet documentation
         */
        _this.keypressEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired preclick event.
         * Use it with `<yaga-tile-layer (preclick)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-preclick Original Leaflet documentation
         */
        _this.preclickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired zoomanim event.
         * Use it with `<yaga-tile-layer (zoomanim)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomanim Original Leaflet documentation
         */
        _this.zoomanimEvent = new core_1.EventEmitter();
        _this.isZooming = false;
        mapProvider.ref = _this;
        layerProvider.ref = _this;
        var moveFn = function () {
            if (_this.isZooming) {
                _this.moveTimeout = setTimeout(moveFn, consts_1.ANIMATION_DELAY);
                return;
            }
            _this.latChange.emit(_this.lat);
            _this.lngChange.emit(_this.lng);
            _this.zoomChange.emit(_this.zoom);
            _this.moveTimeout = undefined;
        };
        _this.setView([0, 0], 0);
        elementRef.nativeElement.setAttribute("class", elementRef.nativeElement.getAttribute("class") + " yaga-map");
        _this.on("move", function () {
            if (_this.moveTimeout) {
                clearTimeout(_this.moveTimeout);
            }
            _this.moveTimeout = setTimeout(moveFn, consts_1.ANIMATION_DELAY);
        });
        _this.on("zoomstart", function () {
            _this.isZooming = true;
        });
        _this.on("zoomend", function () {
            _this.isZooming = false;
            if (_this.moveTimeout) {
                clearTimeout(_this.moveTimeout);
            }
            _this.moveTimeout = setTimeout(moveFn, consts_1.ANIMATION_DELAY);
        });
        _this.on("baselayerchange", function (event) {
            _this.baselayerchangeEvent.emit(event);
        });
        _this.on("overlayadd", function (event) {
            _this.overlayaddEvent.emit(event);
        });
        _this.on("overlayremove", function (event) {
            _this.overlayremoveEvent.emit(event);
        });
        _this.on("layeradd", function (event) {
            _this.layeraddEvent.emit(event);
        });
        _this.on("layerremove", function (event) {
            _this.layerremoveEvent.emit(event);
        });
        _this.on("zoomlevelschange", function (event) {
            _this.zoomlevelschangeEvent.emit(event);
        });
        _this.on("resize", function (event) {
            _this.resizeEvent.emit(event);
        });
        _this.on("unload", function (event) {
            _this.unloadEvent.emit(event);
        });
        _this.on("viewreset", function (event) {
            _this.viewresetEvent.emit(event);
        });
        _this.on("load", function (event) {
            _this.loadEvent.emit(event);
        });
        _this.on("zoomstart", function (event) {
            _this.zoomstartEvent.emit(event);
        });
        _this.on("movestart", function (event) {
            _this.movestartEvent.emit(event);
        });
        _this.on("zoom", function (event) {
            _this.zoomEvent.emit(event);
        });
        _this.on("move", function (event) {
            _this.moveEvent.emit(event);
        });
        _this.on("zoomend", function (event) {
            _this.zoomendEvent.emit(event);
        });
        _this.on("moveend", function (event) {
            _this.moveendEvent.emit(event);
        });
        _this.on("boxzoomstart", function (event) {
            _this.boxzoomstartEvent.emit(event);
        });
        _this.on("boxzoomend", function (event) {
            _this.boxzoomendEvent.emit(event);
        });
        _this.on("popupopen", function (event) {
            _this.popupopenEvent.emit(event);
        });
        _this.on("popupclose", function (event) {
            _this.popupcloseEvent.emit(event);
        });
        _this.on("autopanstart", function (event) {
            _this.autopanstartEvent.emit(event);
        });
        _this.on("tooltipopen", function (event) {
            _this.tooltipopenEvent.emit(event);
        });
        _this.on("tooltipclose", function (event) {
            _this.tooltipcloseEvent.emit(event);
        });
        _this.on("click", function (event) {
            _this.clickEvent.emit(event);
        });
        _this.on("dblclick", function (event) {
            _this.dblclickEvent.emit(event);
        });
        _this.on("mousedown", function (event) {
            _this.mousedownEvent.emit(event);
        });
        _this.on("mouseup", function (event) {
            _this.mouseupEvent.emit(event);
        });
        _this.on("mouseover", function (event) {
            _this.mouseoverEvent.emit(event);
        });
        _this.on("mouseout", function (event) {
            _this.mouseoutEvent.emit(event);
        });
        _this.on("mousemove", function (event) {
            _this.mousemoveEvent.emit(event);
        });
        _this.on("contextmenu", function (event) {
            _this.contextmenuEvent.emit(event);
        });
        _this.on("keypress", function (event) {
            _this.keypressEvent.emit(event);
        });
        _this.on("preclick", function (event) {
            _this.preclickEvent.emit(event);
        });
        _this.on("zoomanim", function (event) {
            _this.zoomanimEvent.emit(event);
        });
        return _this;
    }
    /**
     * This function gets called from Angular after initializing the html-component.
     * @link https://angular.io/docs/ts/latest/api/core/index/AfterViewInit-class.html
     */
    MapComponent.prototype.ngAfterViewInit = function () {
        this.invalidateSize(false);
    };
    Object.defineProperty(MapComponent.prototype, "zoom", {
        get: function () {
            return this.getZoom();
        },
        /*setZoom(zoom: number, options?: ZoomPanOptions): this {
         if (this.zoom === zoom) {
         return;
         }
         this.zoomChange.emit(zoom);
         return super.setZoom(zoom, options)
         }*/
        // already handled with moveend
        // setView(center: LatLngExpression, zoom: number, options?: ZoomPanOptions): this {
        /**
         * Two-Way bound property for the zoom.
         * Use it with `<yaga-map [(zoom)]="someValue">` or `<yaga-map [zoom]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setzoom Original Leaflet documentation
         */
        set: function (val) {
            this.setZoom(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "lat", {
        get: function () {
            return this.getCenter().lat;
        },
        /**
         * Two-Way bound property for the latitude.
         * Use it with `<yaga-map [(lat)]="someValue">` or `<yaga-map [lat]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setview Original Leaflet documentation
         */
        set: function (val) {
            var coords = new leaflet_1.LatLng(val, this.getCenter().lng);
            this.setView(coords, this.zoom);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "lng", {
        get: function () {
            return this.getCenter().lng;
        },
        /**
         * Two-Way bound property for the longitude.
         * Use it with `<yaga-map [(lng)]="someValue">` or `<yaga-map [lng]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setview Original Leaflet documentation
         */
        set: function (val) {
            var coords = new leaflet_1.LatLng(this.getCenter().lat, val);
            this.setView(coords, this.zoom);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Derived method of the original setMinZoom method.
     * @link http://leafletjs.com/reference-1.2.0.html#map-setminzoom Original Leaflet documentation
     */
    MapComponent.prototype.setMinZoom = function (val) {
        this.minZoomChange.emit(val);
        return _super.prototype.setMinZoom.call(this, val);
    };
    Object.defineProperty(MapComponent.prototype, "minZoom", {
        get: function () {
            return this.getMinZoom();
        },
        /**
         * Two-Way bound property for the minimal availabe zoom.
         * Use it with `<yaga-map [(minZoom)]="someValue">` or `<yaga-map [minZoom]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setminzoom Original Leaflet documentation
         */
        set: function (val) {
            this.setMinZoom(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Derived method of the original setMaxZoom method.
     * @link http://leafletjs.com/reference-1.2.0.html#map-setmaxzoom Original Leaflet documentation
     */
    MapComponent.prototype.setMaxZoom = function (val) {
        this.maxZoomChange.emit(val);
        return _super.prototype.setMaxZoom.call(this, val);
    };
    Object.defineProperty(MapComponent.prototype, "maxZoom", {
        get: function () {
            return this.getMaxZoom();
        },
        /**
         * Two-Way bound property for the maximal availabe zoom.
         * Use it with `<yaga-map [(maxZoom)]="someValue">` or `<yaga-map [maxZoom]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-setmaxzoom Original Leaflet documentation
         */
        set: function (val) {
            this.setMaxZoom(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Inherited function to provide event emitting.
     */
    MapComponent.prototype.setMaxBounds = function (bounds) {
        _super.prototype.setMaxBounds.call(this, bounds);
        this.maxBoundsChange.emit(this.maxBounds);
        return this;
    };
    Object.defineProperty(MapComponent.prototype, "crs", {
        get: function () {
            return this.options.crs;
        },
        /**
         * One-Way property for the Coordinate Reference System.
         * Use it with `<yaga-map [crs]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-crs Original Leaflet documentation
         */
        set: function (val) {
            this.options.crs = val;
            var keys = Object.keys(this._layers);
            for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
                var key = keys_1[_i];
                if (typeof this._layers[key].redraw === "function") {
                    this._layers[key].redraw();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "maxBounds", {
        get: function () {
            return this.options.maxBounds;
        },
        /**
         * Two-Way bound property for the maximal bounds.
         * Use it with `<yaga-map [(maxBounds)]="someValue">` or `<yaga-map [maxBounds]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-maxbounds Original Leaflet documentation
         */
        set: function (val) {
            this.setMaxBounds(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "closePopupOnClick", {
        get: function () {
            return this.options.closePopupOnClick;
        },
        // One-way Input
        /**
         * Input for the closePopupOnClick.
         * Use it with `<yaga-map [closePopupOnClick]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-closepopuponclick Original Leaflet documentation
         */
        set: function (val) {
            this.options.closePopupOnClick = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "zoomSnap", {
        get: function () {
            return this.options.zoomSnap;
        },
        /**
         * Input for the zoomSnap.
         * Use it with `<yaga-map [zoomSnap]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomsnap Original Leaflet documentation
         */
        set: function (val) {
            this.options.zoomSnap = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "zoomDelta", {
        get: function () {
            return this.options.zoomDelta;
        },
        /**
         * Input for the zoomDelta.
         * Use it with `<yaga-map [zoomDelta]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomdelta Original Leaflet documentation
         */
        set: function (val) {
            this.options.zoomDelta = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "trackResize", {
        get: function () {
            return this.options.trackResize;
        },
        /**
         * Input for the trackResize.
         * Use it with `<yaga-map [trackResize]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-trackresize Original Leaflet documentation
         */
        set: function (val) {
            this.options.trackResize = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "boxZoomEnabled", {
        get: function () {
            return this.boxZoom.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the boxZoomEnabled.
         * Use it with `<yaga-map [boxZoomEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-boxzoom Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.boxZoom.enable();
                return;
            }
            this.boxZoom.disable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "doubleClickZoomEnabled", {
        get: function () {
            return this.doubleClickZoom.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the doubleClickZoomEnabled.
         * Use it with `<yaga-map [doubleClickZoomEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-doubleclickzoom Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.doubleClickZoom.enable();
                return;
            }
            this.doubleClickZoom.disable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "draggingEnabled", {
        get: function () {
            return this.dragging.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the draggingEnabled.
         * Use it with `<yaga-map [draggingEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-dragging Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.dragging.enable();
                return;
            }
            this.dragging.disable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "fadeAnimation", {
        get: function () {
            return !!this.options.fadeAnimation;
        },
        /**
         * Input for the fadeAnimation.
         * Use it with `<yaga-map [fadeAnimation]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-fadeanimation Original Leaflet documentation
         */
        set: function (val) {
            this.options.fadeAnimation = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "markerZoomAnimation", {
        get: function () {
            return !!this.options.markerZoomAnimation;
        },
        /**
         * Input for the markerZoomAnimation.
         * Use it with `<yaga-map [markerZoomAnimation]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-markerzoomanimation Original Leaflet documentation
         */
        set: function (val) {
            this.options.markerZoomAnimation = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "transform3DLimit", {
        get: function () {
            return this.options.transform3DLimit;
        },
        /**
         * Input for the transform3DLimit.
         * Use it with `<yaga-map [transform3DLimit]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-transform3dlimit Original Leaflet documentation
         */
        set: function (val) {
            this.options.transform3DLimit = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "zoomAnimation", {
        get: function () {
            return !!this.options.zoomAnimation;
        },
        /**
         * Input for the zoomAnimation.
         * Use it with `<yaga-map [zoomAnimation]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomanimation Original Leaflet documentation
         */
        set: function (val) {
            this.options.zoomAnimation = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "zoomAnimationThreshold", {
        get: function () {
            return this.options.zoomAnimationThreshold;
        },
        /**
         * Input for the zoomAnimationThreshold.
         * Use it with `<yaga-map [zoomAnimationThreshold]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-zoomanimationthreshold Original Leaflet documentation
         */
        set: function (val) {
            this.options.zoomAnimationThreshold = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "inertia", {
        get: function () {
            return !!this.options.inertia;
        },
        /**
         * Input for the inertia.
         * Use it with `<yaga-map [inertia]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-inertia Original Leaflet documentation
         */
        set: function (val) {
            this.options.inertia = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "inertiaDeceleration", {
        get: function () {
            return this.options.inertiaDeceleration;
        },
        /**
         * Input for the inertiaDeceleration.
         * Use it with `<yaga-map [inertiaDeceleration]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-inertiadeceleration Original Leaflet documentation
         */
        set: function (val) {
            this.options.inertiaDeceleration = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "inertiaMaxSpeed", {
        get: function () {
            return this.options.inertiaMaxSpeed;
        },
        /**
         * Input for the inertiaMaxSpeed.
         * Use it with `<yaga-map [inertiaMaxSpeed]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-inertiamaxspeed Original Leaflet documentation
         */
        set: function (val) {
            this.options.inertiaMaxSpeed = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "easeLinearity", {
        get: function () {
            return this.options.easeLinearity;
        },
        /**
         * Input for the easeLinearity.
         * Use it with `<yaga-map [easeLinearity]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-easelinearity Original Leaflet documentation
         */
        set: function (val) {
            this.options.easeLinearity = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "worldCopyJump", {
        get: function () {
            return !!this.options.worldCopyJump;
        },
        /**
         * Input for the worldCopyJump.
         * Use it with `<yaga-map [worldCopyJump]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-worldcopyjump Original Leaflet documentation
         */
        set: function (val) {
            this.options.worldCopyJump = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "maxBoundsViscosity", {
        get: function () {
            return this.options.maxBoundsViscosity;
        },
        /**
         * Input for the maxBoundsViscosity.
         * Use it with `<yaga-map [maxBoundsViscosity]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-maxboundsviscosity Original Leaflet documentation
         */
        set: function (val) {
            this.options.maxBoundsViscosity = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "keyboardEnabled", {
        get: function () {
            return this.keyboard.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the keyboardEnabled.
         * Use it with `<yaga-map [keyboardEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-keyboard Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.keyboard.enable();
                return;
            }
            this.keyboard.disable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "keyboardPanDelta", {
        get: function () {
            return this.options.keyboardPanDelta;
        },
        /**
         * Input for the keyboardPanDelta.
         * Use it with `<yaga-map [keyboardPanDelta]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-keyboardpandelta Original Leaflet documentation
         */
        set: function (val) {
            this.options.keyboardPanDelta = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "scrollWheelZoomEnabled", {
        get: function () {
            return this.scrollWheelZoom.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the scrollWheelZoomEnabled.
         * Use it with `<yaga-map [scrollWheelZoomEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-scrollwheelzoom Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.scrollWheelZoom.enable();
                return;
            }
            this.scrollWheelZoom.disable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "wheelDebounceTime", {
        get: function () {
            return this.options.wheelDebounceTime;
        },
        /**
         * Input for the wheelDebounceTime.
         * Use it with `<yaga-map [wheelDebounceTime]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-wheeldebouncetime Original Leaflet documentation
         */
        set: function (val) {
            this.options.wheelDebounceTime = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "wheelPxPerZoomLevel", {
        get: function () {
            return this.options.wheelPxPerZoomLevel;
        },
        /**
         * Input for the wheelPxPerZoomLevel.
         * Use it with `<yaga-map [wheelPxPerZoomLevel]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-wheelpxperzoomlevel Original Leaflet documentation
         */
        set: function (val) {
            this.options.wheelPxPerZoomLevel = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "tapEnabled", {
        get: function () {
            return !!this.options.tap;
        },
        /**
         * Input for the tapEnabled.
         * Use it with `<yaga-map [tapEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-tap Original Leaflet documentation
         */
        set: function (val) {
            this.options.tap = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "tapTolerance", {
        get: function () {
            return this.options.tapTolerance;
        },
        /**
         * Input for the tapTolerance.
         * Use it with `<yaga-map [tapTolerance]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-taptolerance Original Leaflet documentation
         */
        set: function (val) {
            this.options.tapTolerance = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "bounceAtZoomLimits", {
        get: function () {
            return !!this.options.bounceAtZoomLimits;
        },
        /**
         * Input for the bounceAtZoomLimits.
         * Use it with `<yaga-map [bounceAtZoomLimits]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-bounceatzoomlimits Original Leaflet documentation
         */
        set: function (val) {
            this.options.bounceAtZoomLimits = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapComponent.prototype, "touchZoomEnabled", {
        get: function () {
            return this.touchZoom.enabled();
        },
        // maybe 2way!?!
        /**
         * Input for the touchZoomEnabled.
         * Use it with `<yaga-map [touchZoomEnabled]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#map-touchzoom Original Leaflet documentation
         */
        set: function (val) {
            if (val) {
                this.touchZoom.enable();
                return;
            }
            this.touchZoom.disable();
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "latChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "lngChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "minZoomChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "maxZoomChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "maxBoundsChange", void 0);
    __decorate([
        core_1.Output("baselayerchange"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "baselayerchangeEvent", void 0);
    __decorate([
        core_1.Output("overlayadd"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "overlayaddEvent", void 0);
    __decorate([
        core_1.Output("overlayremove"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "overlayremoveEvent", void 0);
    __decorate([
        core_1.Output("layeradd"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "layeraddEvent", void 0);
    __decorate([
        core_1.Output("layerremove"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "layerremoveEvent", void 0);
    __decorate([
        core_1.Output("zoomlevelschange"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomlevelschangeEvent", void 0);
    __decorate([
        core_1.Output("resize"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "resizeEvent", void 0);
    __decorate([
        core_1.Output("unload"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "unloadEvent", void 0);
    __decorate([
        core_1.Output("viewreset"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "viewresetEvent", void 0);
    __decorate([
        core_1.Output("load"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "loadEvent", void 0);
    __decorate([
        core_1.Output("zoomstart"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomstartEvent", void 0);
    __decorate([
        core_1.Output("movestart"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "movestartEvent", void 0);
    __decorate([
        core_1.Output("zoom"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomEvent", void 0);
    __decorate([
        core_1.Output("move"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "moveEvent", void 0);
    __decorate([
        core_1.Output("zoomend"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomendEvent", void 0);
    __decorate([
        core_1.Output("moveend"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "moveendEvent", void 0);
    __decorate([
        core_1.Output("boxzoomstart"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "boxzoomstartEvent", void 0);
    __decorate([
        core_1.Output("boxzoomend"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "boxzoomendEvent", void 0);
    __decorate([
        core_1.Output("popupopen"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "popupopenEvent", void 0);
    __decorate([
        core_1.Output("popupclose"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "popupcloseEvent", void 0);
    __decorate([
        core_1.Output("autopanstart"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "autopanstartEvent", void 0);
    __decorate([
        core_1.Output("tooltipopen"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "tooltipopenEvent", void 0);
    __decorate([
        core_1.Output("tooltipclose"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "tooltipcloseEvent", void 0);
    __decorate([
        core_1.Output("click"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "clickEvent", void 0);
    __decorate([
        core_1.Output("dblclick"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "dblclickEvent", void 0);
    __decorate([
        core_1.Output("mousedown"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "mousedownEvent", void 0);
    __decorate([
        core_1.Output("mouseup"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "mouseupEvent", void 0);
    __decorate([
        core_1.Output("mouseover"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "mouseoverEvent", void 0);
    __decorate([
        core_1.Output("mouseout"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "mouseoutEvent", void 0);
    __decorate([
        core_1.Output("mousemove"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "mousemoveEvent", void 0);
    __decorate([
        core_1.Output("contextmenu"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "contextmenuEvent", void 0);
    __decorate([
        core_1.Output("keypress"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "keypressEvent", void 0);
    __decorate([
        core_1.Output("preclick"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "preclickEvent", void 0);
    __decorate([
        core_1.Output("zoomanim"),
        __metadata("design:type", core_1.EventEmitter)
    ], MapComponent.prototype, "zoomanimEvent", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], MapComponent.prototype, "zoom", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], MapComponent.prototype, "lat", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], MapComponent.prototype, "lng", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], MapComponent.prototype, "minZoom", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], MapComponent.prototype, "maxZoom", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "crs", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", leaflet_1.LatLngBounds),
        __metadata("design:paramtypes", [leaflet_1.LatLngBounds])
    ], MapComponent.prototype, "maxBounds", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "closePopupOnClick", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "zoomSnap", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "zoomDelta", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "trackResize", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "boxZoomEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "doubleClickZoomEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "draggingEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "fadeAnimation", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "markerZoomAnimation", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "transform3DLimit", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "zoomAnimation", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "zoomAnimationThreshold", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "inertia", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "inertiaDeceleration", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "inertiaMaxSpeed", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "easeLinearity", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "worldCopyJump", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "maxBoundsViscosity", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "keyboardEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "keyboardPanDelta", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "scrollWheelZoomEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "wheelDebounceTime", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "wheelPxPerZoomLevel", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "tapEnabled", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], MapComponent.prototype, "tapTolerance", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "bounceAtZoomLimits", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], MapComponent.prototype, "touchZoomEnabled", null);
    MapComponent = __decorate([
        core_1.Component({
            providers: [layer_group_provider_1.LayerGroupProvider, map_provider_1.MapProvider],
            selector: "yaga-map",
            styles: [":host { display: block; }"],
            template: "<span style=\"display: none\"><ng-content></ng-content></span>",
        }),
        __param(0, core_1.Inject(core_1.ElementRef)),
        __param(1, core_1.Host()),
        __metadata("design:paramtypes", [core_1.ElementRef,
            layer_group_provider_1.LayerGroupProvider,
            map_provider_1.MapProvider])
    ], MapComponent);
    return MapComponent;
}(leaflet_1.Map));
exports.MapComponent = MapComponent;
