// @yaga/leaflet-ng2/lib/scale-control.directive.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var leaflet_1 = require("leaflet");
var map_provider_1 = require("./map.provider");
var mouse_event_helper_1 = require("./mouse-event-helper");
/**
 * Angular2 directive for the attribution-control of Leaflet.
 *
 * *You can use this directive in an Angular2 template after importing `YagaModule`.*
 *
 * How to use in a template:
 * ```html
 * <yaga-map>
 *     <yaga-scale-control
 *         [(display)]="..."
 *         [(zIndex)]="..."
 *         [(position)]="..."
 *
 *         [metric]="..."
 *         [imperial]="..."
 *         [maxWidth]="..."
 *
 *         (add)="..."
 *         (remove)="..."
 *         (click)="..."
 *         (dblclick)="..."
 *         (mousedown)="..."
 *         (mouseover)="..."
 *         (mouseout)="..."
 *         >
 *     </yaga-scale-control>
 * </yaga-map>
 * ```
 *
 * @link http://leafletjs.com/reference-1.2.0.html#control-scale Original Leaflet documentation
 * @link https://leaflet-ng2.yagajs.org/latest/browser-test?grep=Scale-Control%20Directive Unit-Test
 * @link https://leaflet-ng2.yagajs.org/latest/coverage/lcov-report/lib/attribution-control.directive.js.html
 * Test coverage
 * @link https://leaflet-ng2.yagajs.org/latest/typedoc/classes/scalecontroldirective.html API documentation
 * @example https://leaflet-ng2.yagajs.org/latest/examples/scale-control-directive/
 */
var ScaleControlDirective = /** @class */ (function (_super) {
    __extends(ScaleControlDirective, _super);
    function ScaleControlDirective(mapProvider) {
        var _this = _super.call(this) || this;
        _this.mapProvider = mapProvider;
        /**
         * Two-Way bound property for the display status of the control.
         * Use it with `<yaga-scale-control [(display)]="someValue">`
         * or `<yaga-scale-control (displayChange)="processEvent($event)">`
         */
        _this.displayChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the zIndex of the control.
         * Use it with `<yaga-scale-control [(zIndex)]="someValue">`
         * or `<yaga-scale-control (zIndexChange)="processEvent($event)">`
         */
        _this.zIndexChange = new core_1.EventEmitter();
        /**
         * Two-Way bound property for the position of the control.
         * Use it with `<yaga-scale-control [(position)]="someValue">`
         * or `<yaga-scale-control (positionChange)="processEvent($event)">`
         */
        _this.positionChange = new core_1.EventEmitter();
        /**
         * From leaflet fired add event.
         * Use it with `<yaga-scale-control (add)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-add Original Leaflet documentation
         */
        _this.addEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired remove event.
         * Use it with `<yaga-scale-control (remove)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-remove Original Leaflet documentation
         */
        _this.removeEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired click event.
         * Use it with `<yaga-scale-control (click)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-click Original Leaflet documentation
         */
        _this.clickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired dblclick event.
         * Use it with `<yaga-scale-control (dblclick)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-dblclick Original Leaflet documentation
         */
        _this.dblclickEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mousedown event.
         * Use it with `<yaga-scale-control (mousedown)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-mousedown Original Leaflet documentation
         */
        _this.mousedownEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseover event.
         * Use it with `<yaga-scale-control (mouseover)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-mouseover Original Leaflet documentation
         */
        _this.mouseoverEvent = new core_1.EventEmitter();
        /**
         * From leaflet fired mouseout event.
         * Use it with `<yaga-scale-control (mouseout)="processEvent($event)">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-mouseout Original Leaflet documentation
         */
        _this.mouseoutEvent = new core_1.EventEmitter();
        _this.mapProvider.ref.addControl(_this);
        // Events
        _this.getContainer().addEventListener("click", function (event) {
            _this.clickEvent.emit(mouse_event_helper_1.enhanceMouseEvent(event, _this._map));
        });
        _this.getContainer().addEventListener("dblclick", function (event) {
            _this.dblclickEvent.emit(mouse_event_helper_1.enhanceMouseEvent(event, _this._map));
        });
        _this.getContainer().addEventListener("mousedown", function (event) {
            _this.mousedownEvent.emit(mouse_event_helper_1.enhanceMouseEvent(event, _this._map));
        });
        _this.getContainer().addEventListener("mouseover", function (event) {
            _this.mouseoverEvent.emit(mouse_event_helper_1.enhanceMouseEvent(event, _this._map));
        });
        _this.getContainer().addEventListener("mouseout", function (event) {
            _this.mouseoutEvent.emit(mouse_event_helper_1.enhanceMouseEvent(event, _this._map));
        });
        return _this;
    }
    /**
     * Internal method to provide the removal of the control in Leaflet, when removing it from the Angular template
     */
    ScaleControlDirective.prototype.ngOnDestroy = function () {
        this.mapProvider.ref.removeControl(this);
    };
    /**
     * Derived remove function
     */
    ScaleControlDirective.prototype.remove = function () {
        /* tslint:disable */
        _super.prototype.remove.call(this);
        this.displayChange.emit(false);
        this.removeEvent.emit({ target: this, type: "remove" });
        return this;
    };
    /**
     * Derived addTo function
     */
    ScaleControlDirective.prototype.addTo = function (map) {
        /* tslint:disable */
        _super.prototype.addTo.call(this, map);
        this.displayChange.emit(true);
        this.addEvent.emit({ target: this, type: "add" });
        return this;
    };
    /**
     * Derived method of the original setPosition.
     * @link http://leafletjs.com/reference-1.2.0.html#control-scale-setposition Original Leaflet documentation
     */
    ScaleControlDirective.prototype.setPosition = function (val) {
        _super.prototype.setPosition.call(this, val);
        this.positionChange.emit(val);
        return this;
    };
    Object.defineProperty(ScaleControlDirective.prototype, "opacity", {
        get: function () {
            var val = this.getContainer().style.opacity;
            if (typeof val === "string") {
                return parseFloat(val);
            }
        },
        /**
         * Two-Way bound property for the opacity.
         * Use it with `<yaga-scale-control [(opacity)]="someValue">`
         * or `<yaga-scale-control [opacity]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-opacity Original Leaflet documentation
         */
        set: function (val) {
            if (typeof val === "number") {
                this.getContainer().style.opacity = val.toString();
                return;
            }
            this.getContainer().style.opacity = null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "display", {
        get: function () {
            return !!this._map && this.getContainer().style.display !== "none";
        },
        /**
         * Two-Way bound property for the display state.
         * Use it with `<yaga-scale-control [(display)]="someValue">`
         * or `<yaga-scale-control [display]="someValue">`
         */
        set: function (val) {
            if (!this._map) {
                // No map available...
                return;
            }
            if (val) {
                this.getContainer().style.display = "";
                return;
            }
            this.getContainer().style.display = "none";
            return;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "position", {
        get: function () {
            return this.getPosition();
        },
        /**
         * Two-Way bound property for the position.
         * Use it with `<yaga-scale-control [(position)]="someValue">`
         * or `<yaga-scale-control [position]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-position Original Leaflet documentation
         */
        set: function (val) {
            this.setPosition(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "zIndex", {
        get: function () {
            var val = this.getContainer().style.zIndex;
            if (typeof val === "string") {
                return parseInt(val, 10);
            }
        },
        /**
         * Two-Way bound property for the zIndex of the control.
         * Use it with `<yaga-scale-control [(zIndex)]="someValue">`
         * or `<yaga-scale-control (zIndexChange)="processEvent($event)">`
         */
        set: function (zIndex) {
            if (typeof zIndex === "number") {
                this.getContainer().style.zIndex = zIndex.toString();
                return;
            }
            this.getContainer().style.zIndex = null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "maxWidth", {
        get: function () {
            return this.options.maxWidth;
        },
        /**
         * Input for scale max-width.
         * Use it with `<yaga-scale-control [(maxWidth)]="someValue">`
         * or `<yaga-scale-control [maxWidth]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-maxwidth Original Leaflet documentation
         */
        set: function (val) {
            this.options.maxWidth = val;
            this._update();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "metric", {
        get: function () {
            return !!this.options.metric;
        },
        /**
         * Input for state of metric-scale state.
         * Use it with `<yaga-scale-control [(metric)]="someValue">`
         * or `<yaga-scale-control [metric]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-metric Original Leaflet documentation
         */
        set: function (val) {
            while (this.getContainer().hasChildNodes()) {
                this.getContainer().removeChild(this.getContainer().lastChild);
            }
            this.options.metric = val;
            this._addScales(this.options, "leaflet-control-scale-line", this.getContainer());
            this._update();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScaleControlDirective.prototype, "imperial", {
        get: function () {
            return !!this.options.imperial;
        },
        /**
         * Input for state of imperial-scale state.
         * Use it with `<yaga-scale-control [(imperial)]="someValue">`
         * or `<yaga-scale-control [imperial]="someValue">`
         * @link http://leafletjs.com/reference-1.2.0.html#control-scale-imperial Original Leaflet documentation
         */
        set: function (val) {
            while (this.getContainer().hasChildNodes()) {
                this.getContainer().removeChild(this.getContainer().lastChild);
            }
            this.options.imperial = val;
            this._addScales(this.options, "leaflet-control-scale-line", this.getContainer());
            this._update();
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "displayChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "zIndexChange", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "positionChange", void 0);
    __decorate([
        core_1.Output("add"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "addEvent", void 0);
    __decorate([
        core_1.Output("remove"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "removeEvent", void 0);
    __decorate([
        core_1.Output("click"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "clickEvent", void 0);
    __decorate([
        core_1.Output("dblclick"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "dblclickEvent", void 0);
    __decorate([
        core_1.Output("mousedown"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "mousedownEvent", void 0);
    __decorate([
        core_1.Output("mouseover"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "mouseoverEvent", void 0);
    __decorate([
        core_1.Output("mouseout"),
        __metadata("design:type", core_1.EventEmitter)
    ], ScaleControlDirective.prototype, "mouseoutEvent", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], ScaleControlDirective.prototype, "opacity", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ScaleControlDirective.prototype, "display", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ScaleControlDirective.prototype, "position", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], ScaleControlDirective.prototype, "zIndex", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], ScaleControlDirective.prototype, "maxWidth", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ScaleControlDirective.prototype, "metric", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], ScaleControlDirective.prototype, "imperial", null);
    ScaleControlDirective = __decorate([
        core_1.Directive({
            selector: "yaga-scale-control",
        }),
        __metadata("design:paramtypes", [map_provider_1.MapProvider])
    ], ScaleControlDirective);
    return ScaleControlDirective;
}(leaflet_1.Control.Scale));
exports.ScaleControlDirective = ScaleControlDirective;
