// @yaga/leaflet-ng2/__ivy_ngcc__/lib/consts.js
"use strict";
/* tslint:disable:max-line-length */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Base64 data-url for a transparent 1px png image
 */
exports.TRANSPARENT_PIXEL = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=";
/**
 * Default value for delays caused by animations (in milliseconds)
 */
exports.ANIMATION_DELAY = 50;
/**
 * Some lorem ipsum test content
 */
exports.EXAMPLE_CONTENT = "Vel ipsum odit quia velit omnis illo voluptatem ut. Aperiam porro voluptates maiores.";
/**
 * URL-schema of the official osm tile-server
 */
exports.OSM_TILE_LAYER_URL = "http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png";
/**
 * URL for an example image for image-overlays
 */
exports.IMAGE_OVERLAY_URL = "http://download.osgeo.org/livedvd/10.0/desktop10_osmF32.png";
/**
 * Base URL for a sample WMS service
 */
exports.EXAMPLE_WMS_LAYER_URL = "http://www.wms.nrw.de/geobasis/wms_nw_dtk?";
/**
 * Layers of the example WMS service
 */
exports.EXAMPLE_WMS_LAYER_NAMES = ["nw_dtk_col"];
/**
 * Default attribution prefix
 */
exports.ATTRIBUTION_PREFIX = "<a href=\"https://yagajs.org\" title=\"YAGA\">YAGA</a> | <a href=\"https://leaflet-ng2.yagajs.org\" title=\"Leaflet in Angular2\">leaflet-ng2</a>";
/**
 * Empty default style
 */
exports.DEFAULT_STYLE = {};
