// @yaga/leaflet-ng2/__ivy_ngcc__/lib/yaga.module.js
"use strict";
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./attribution-control.directive');
var ɵngcc2 = require('./circle-marker.directive');
var ɵngcc3 = require('./circle.directive');
var ɵngcc4 = require('./div-icon.directive');
var ɵngcc5 = require('./feature-group.directive');
var ɵngcc6 = require('./geojson.directive');
var ɵngcc7 = require('./icon.directive');
var ɵngcc8 = require('./image-overlay.directive');
var ɵngcc9 = require('./layer-group.directive');
var ɵngcc10 = require('./layers-control.directive');
var ɵngcc11 = require('./map.component');
var ɵngcc12 = require('./marker.directive');
var ɵngcc13 = require('./overlay-layer.directive');
var ɵngcc14 = require('./polygon.directive');
var ɵngcc15 = require('./polyline.directive');
var ɵngcc16 = require('./popup.directive');
var ɵngcc17 = require('./rectangle.directive');
var ɵngcc18 = require('./scale-control.directive');
var ɵngcc19 = require('./tile-layer.directive');
var ɵngcc20 = require('./tooltip.directive');
var ɵngcc21 = require('./wms-layer.directive');
var ɵngcc22 = require('./zoom-control.directive');
var ɵngcc23 = require('./base-layer.directive');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var attribution_control_directive_1 = require("./attribution-control.directive");
var base_layer_directive_1 = require("./base-layer.directive");
var circle_marker_directive_1 = require("./circle-marker.directive");
var circle_directive_1 = require("./circle.directive");
var div_icon_directive_1 = require("./div-icon.directive");
var feature_group_directive_1 = require("./feature-group.directive");
var geojson_directive_1 = require("./geojson.directive");
var icon_directive_1 = require("./icon.directive");
var image_overlay_directive_1 = require("./image-overlay.directive");
var layer_group_directive_1 = require("./layer-group.directive");
var layers_control_directive_1 = require("./layers-control.directive");
var map_component_1 = require("./map.component");
var marker_directive_1 = require("./marker.directive");
var overlay_layer_directive_1 = require("./overlay-layer.directive");
var polygon_directive_1 = require("./polygon.directive");
var polyline_directive_1 = require("./polyline.directive");
var popup_directive_1 = require("./popup.directive");
var rectangle_directive_1 = require("./rectangle.directive");
var scale_control_directive_1 = require("./scale-control.directive");
var tile_layer_directive_1 = require("./tile-layer.directive");
var tooltip_directive_1 = require("./tooltip.directive");
var wms_layer_directive_1 = require("./wms-layer.directive");
var zoom_control_directive_1 = require("./zoom-control.directive");
/**
 * Angular 2++ module that you can use to import all directives of YAGA"s leaflet-ng2 into your Angular application.
 *
 * ```
 * import { NgModule } from "@angular/core";
 * import { YagaModule } from "@yaga/leaflet-ng2";
 *
 * ; @NgModule({
 *     imports: [
 *         YagaModule,
 *     ],
 * })
 * export class AppModule {}
 * ```
 */
var YagaModule = /** @class */ (function () {
    function YagaModule() {
    }
YagaModule.ɵfac = function YagaModule_Factory(t) { return new (t || YagaModule)(); };
YagaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: YagaModule });
YagaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(YagaModule, { declarations: [ɵngcc1.AttributionControlDirective, ɵngcc2.CircleMarkerDirective, ɵngcc3.CircleDirective, ɵngcc4.DivIconDirective, ɵngcc5.FeatureGroupDirective, ɵngcc6.GeoJSONDirective, ɵngcc7.IconDirective, ɵngcc8.ImageOverlayDirective, ɵngcc9.LayerGroupDirective, ɵngcc10.LayersControlDirective, ɵngcc11.MapComponent, ɵngcc12.MarkerDirective, ɵngcc13.OverlayLayerDirective, ɵngcc14.PolygonDirective, ɵngcc15.PolylineDirective, ɵngcc16.PopupDirective, ɵngcc17.RectangleDirective, ɵngcc18.ScaleControlDirective, ɵngcc19.TileLayerDirective, ɵngcc20.TooltipDirective, ɵngcc21.WmsLayerDirective, ɵngcc22.ZoomControlDirective, ɵngcc23.BaseLayerDirective], exports: [ɵngcc1.AttributionControlDirective, ɵngcc2.CircleMarkerDirective, ɵngcc3.CircleDirective, ɵngcc4.DivIconDirective, ɵngcc5.FeatureGroupDirective, ɵngcc6.GeoJSONDirective, ɵngcc7.IconDirective, ɵngcc8.ImageOverlayDirective, ɵngcc9.LayerGroupDirective, ɵngcc10.LayersControlDirective, ɵngcc11.MapComponent, ɵngcc12.MarkerDirective, ɵngcc13.OverlayLayerDirective, ɵngcc14.PolygonDirective, ɵngcc15.PolylineDirective, ɵngcc16.PopupDirective, ɵngcc17.RectangleDirective, ɵngcc18.ScaleControlDirective, ɵngcc19.TileLayerDirective, ɵngcc20.TooltipDirective, ɵngcc21.WmsLayerDirective, ɵngcc22.ZoomControlDirective, ɵngcc23.BaseLayerDirective] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(YagaModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [
                    attribution_control_directive_1.AttributionControlDirective,
                    circle_marker_directive_1.CircleMarkerDirective,
                    circle_directive_1.CircleDirective,
                    div_icon_directive_1.DivIconDirective,
                    feature_group_directive_1.FeatureGroupDirective,
                    geojson_directive_1.GeoJSONDirective,
                    icon_directive_1.IconDirective,
                    image_overlay_directive_1.ImageOverlayDirective,
                    layer_group_directive_1.LayerGroupDirective,
                    layers_control_directive_1.LayersControlDirective,
                    map_component_1.MapComponent,
                    marker_directive_1.MarkerDirective,
                    overlay_layer_directive_1.OverlayLayerDirective,
                    polygon_directive_1.PolygonDirective,
                    polyline_directive_1.PolylineDirective,
                    popup_directive_1.PopupDirective,
                    rectangle_directive_1.RectangleDirective,
                    scale_control_directive_1.ScaleControlDirective,
                    tile_layer_directive_1.TileLayerDirective,
                    tooltip_directive_1.TooltipDirective,
                    wms_layer_directive_1.WmsLayerDirective,
                    zoom_control_directive_1.ZoomControlDirective,
                    base_layer_directive_1.BaseLayerDirective,
                ],
                exports: [
                    attribution_control_directive_1.AttributionControlDirective,
                    circle_marker_directive_1.CircleMarkerDirective,
                    circle_directive_1.CircleDirective,
                    div_icon_directive_1.DivIconDirective,
                    feature_group_directive_1.FeatureGroupDirective,
                    geojson_directive_1.GeoJSONDirective,
                    icon_directive_1.IconDirective,
                    image_overlay_directive_1.ImageOverlayDirective,
                    layer_group_directive_1.LayerGroupDirective,
                    layers_control_directive_1.LayersControlDirective,
                    map_component_1.MapComponent,
                    marker_directive_1.MarkerDirective,
                    overlay_layer_directive_1.OverlayLayerDirective,
                    polygon_directive_1.PolygonDirective,
                    polyline_directive_1.PolylineDirective,
                    popup_directive_1.PopupDirective,
                    rectangle_directive_1.RectangleDirective,
                    scale_control_directive_1.ScaleControlDirective,
                    tile_layer_directive_1.TileLayerDirective,
                    tooltip_directive_1.TooltipDirective,
                    wms_layer_directive_1.WmsLayerDirective,
                    zoom_control_directive_1.ZoomControlDirective,
                    base_layer_directive_1.BaseLayerDirective,
                ]
            }]
    }], function () { return []; }, null); })();
    return YagaModule;
}());
exports.YagaModule = YagaModule;
